File: src/main/scala/net/karlv/flang/parser/Lexicon.scala
package net.karlv.flang.parser

import scala.util.parsing.combinator.RegexParsers
import java.io.InputStreamReader
import java.io.InputStream
import scala.util.parsing.input.CharSequenceReader
import scala.util.parsing.combinator.lexical.Lexical
import scala.util.parsing.input.PagedSeqReader
import scala.util.parsing.input.Reader
import scala.collection.immutable.PagedSeq

object Lexicon extends Lexical with RegexParsers {
  
  override type Elem = Char;
  
  def apply(is: InputStream): Reader[Token] =
    new Scanner(new PagedSeqReader(new PagedSeq((a, b, c) => new InputStreamReader(is).read(a, b, c))));
  
  override def whitespace: Parser[Unit] = comment | ws;
  
  override def token: Parser[Token] =
    foldLeft1[Parser[Token]](_ | _)(
      keywords ++ List(id, exprOp, int, float, string, char));
  
  def foldLeft1[A](op: (A, A) => A)(xs: Iterable[A]) = xs.tail.foldLeft(xs.head)(op);

  def keywords: List[Parser[TKeyword]] = List(
      "type", "val", "data", "sig", "(", ")", "open", "closed", "except", "only", "is",
      "rec", "let", "fn", "case", "begin", "do", "?", "where", "end", "in", "of",
      "module", "with", ":", "<:", ":>", "->", "<-", ";", "*", ".", "forall", "infix",
      "left", "right").map(_ ^^ TKeyword);

  def rawId: String = """[a-zA-Z_][a-zA-Z0-9_]*""";
  
  def id: Parser[TId] = rawId.r ^^ TId;

  def exprOp: Parser[TExprOp] = """(\+|-|\*|/|=|<|>)+""".r ^^ TExprOp;

  def int: Parser[TInt] = """[0-9]+""".r ^^ TInt;

  def float: Parser[TFloat] = ("""([0-9]+\.[0-9]*|\.[0-9]+|[0-9]+)""" + exponent).r ^^ TFloat;

  def exponent = """[eE][+\-]?[0-9]+""";

  // TODO better comment syntax, with nestable block comments
  def comment: Parser[Unit] = """//[^\n\r]*\r?\n""".r ~> success(())

  def ws: Parser[Unit] = """[ \t\r\n]*""".r ~> success(());

  def charContent(quote: String, card: String) =
    ("""(\\[btnfr"'\\]|\\u[0-9a-fA-F]{4}|\\[0-3]?[0-7]{1,2}|[^\\""" + quote + "])" + card).r;

  def string: Parser[TString] = "\"" ~> charContent("\"", "*") <~ "\"" ^^ TString;

  def char: Parser[TChar] = "'" ~> charContent("'", "") <~ "'" ^^ TChar;

}

File: hs/Lexicon.hs

module Lexicon where

#include "Common.h"
import Text.Parsec
import Text.Parsec.Char
import Text.Parsec.Combinator
import Text.Parsec.Text
import Token

tokenize :: Text -> Text -> [Token]
tokenize name xs = case parse file (Text.unpack name) xs of
  Left err -> error . show $ err
  Right ts -> ts

file :: Parser [Token]
file = skippable >> many1 tok `sepEndBy` skippable >>= return . concat

skippable = many $ comment <|> spaces

tok = foldl1 (<|>) $ keywords ++ [ident, exprOp, litInt, litFloat, litString, litChar]

keywords =
  map ((>>= return . TKeyword . Text.pack) . string . Text.unpack)
    [ "type", "val", "data", "sig", "(", ")", "open", "closed", "except", "only", "is"
    , "rec", "let", "fn", "case", "begin", "do", "?", "where", "end", "in", "of"
    , "module", "with", ":", "<:", ":>", "->", "<-", ";", "*", ".", "forall", "infix"
    , "left", "right"
    ]

ident = do
  head <- letter <|> char '_'
  tail <- many $ alphaNum <|> char '_'
  return . TId . Text.cons head . Text.pack $ tail

exprOp = many1 (oneOf "+-*/=<>") >>= return . TExprOp . Text.pack

litInt = many1 digit >>= return . TInt . read . Text.pack

litFloat :: Parser Token
litFloat = do
  (intPart, fracPart) <- withIntPart <|> withoutIntPart
  exp <- optionMaybe $ do
    oneOf "eE"
    sign <- optionMaybe $ oneOf "+-"
    let signVal = if sign == Just '-' then -1 else 1
    digs <- many1 digit
    return $ (signVal * read (Text.pack digs) :: Integer)
  let intVal = read (Text.pack intPart) :: Integer
  let fracVal = (read (Text.pack fracPart) :: Integer) % (10 ^ Text.length (Text.pack fracPart))
  return . TFloat $ (fromInteger intVal + fracVal) * (10 ^ maybe 0 id exp)
  where
    withIntPart = do
      intPart <- many1 digit
      char '.'
      fracPart <- many digit
      return (intPart, fracPart)
    withoutIntPart = do
      char '.'
      fracPart <- many1 digit
      return ("0", fracPart)

comment = undefined

litString = undefined

litChar = undefined

{--

  // TODO better comment syntax, with nestable block comments
  def comment: Parser[Unit] = """//[^\n\r]*\r?\n""".r ~> success(())

  def charContent(quote: String, card: String) =
    ("""(\\[btnfr"'\\]|\\u[0-9a-fA-F]{4}|\\[0-3]?[0-7]{1,2}|[^\\""" + quote + "])" + card).r;

  def string: Parser[TString] = "\"" ~> charContent("\"", "*") <~ "\"" ^^ TString;

  def char: Parser[TChar] = "'" ~> charContent("'", "") <~ "'" ^^ TChar;

--}

